death-causes-app/src/components/Calculations/CalculationService.ts

```diff
@@ -0,0 +1,227 @@
+    }
+
+    calculateInnerProbabilities(factorAnswersSubmitted: FactorAnswers, deathcause: DeathCause): DataRow {
+        let uStar = this.calculateUStar(factorAnswersSubmitted, deathcause);
+        let firstOrderDecomposition = this.calculateFirstOrderDecomposition(factorAnswersSubmitted, deathcause);
+
+        let ratio
+        if (uStar.propForDeathcause === 0) {
+            ratio = 0
+        } else {
+            ratio = (uStar.propForDeathcause - uStar.minProbForDeathcause) / uStar.propForDeathcause
+        }
+        let innerCauses: ProbabilityKeyValue = {}
+        for (let key of Object.keys(firstOrderDecomposition)) {
+            innerCauses[key] = firstOrderDecomposition[key] * ratio
+        }
+
+        let res = {
+            name: deathcause.deathCauseName,
+            totalProb: uStar.propForDeathcause,
+            innerCauses: innerCauses
+        }
+
+        return res
+    }
+
+    private calculateUStar(factorAnswersSubmitted: FactorAnswers, deathcause: DeathCause): UStar {
+        let probForDeathcause = this.calculateProbabilityForSingleCauseAndAge(factorAnswersSubmitted, factorAnswersSubmitted['Age'] as number, deathcause)
+        let minimumFactorIntervals: MinimumFactorInputs = {}
+        deathcause.riskFactorGroups.forEach(rfg => {
+            rfg.riskRatioTables.forEach(rrt => {
+                minimumFactorIntervals = { ...minimumFactorIntervals, ...rrt.getMinimumRRFactors() }
+            })
+        })
+
+        let minimumFactorInputs: FactorAnswers = {}
+        for (var key in minimumFactorIntervals) {
+            minimumFactorInputs[key] = minimumFactorIntervals[key].getValueInCell()
+        }
+
+        let minProbForDeathcause = this.calculateProbabilityForSingleCauseAndAge(minimumFactorInputs, factorAnswersSubmitted['Age'] as number, deathcause)
+
+        return { propForDeathcause: probForDeathcause, minProbForDeathcause: minProbForDeathcause, minFactorIntervals: minimumFactorIntervals, minFactorValues: minimumFactorInputs }
+    }
+
+    private calculateFirstOrderDecomposition(factorAnswersSubmitted: FactorAnswers, deathcause: DeathCause): ProbabilityKeyValue {
+        let res: ProbabilityKeyValue = {}
+        for (let rfg of deathcause.riskFactorGroups) {
+            let rfgRes = this.calculateFirstOrderDecompositionForRiskFactorGroup(factorAnswersSubmitted, rfg);
+            res = { ...res, ...rfgRes };
+        }
+
+        let sum = 0;
+        for (let key of Object.keys(res)) {
+            sum += res[key];
+        }
+        for (let key of Object.keys(res)) {
+            if (sum === 0) {
+                res[key] = 1 / Object.keys(res).length
+            } else {
+                res[key] = res[key] / sum;
+            }
+        }
+
+        return res;
+    }
+
+    private calculateFirstOrderDecompositionForRiskFactorGroup(factorAnswersSubmitted: FactorAnswers, riskFactorGroup: RiskFactorGroup) {
+        let res: MinimumRiskRatios = {}
```

Comment:
I stedet for at have dictionaries, hvor keys'ne bliver overskrevet, tror jeg det er bedre at bruge lister, som først senere skrives sammen til dictionaries. Jeg tænker at denne funktion burde returnere noget a la [{factor: "Smoking", direction: "too much", value: 0.4 }, {factor: "Smoking", direction: "too much", value: 0.3}, ...]